## Exploring Hacker News Posts

Find out titles that begin with either Ask HN or Show HN

In [2]:
from csv import reader
with open(r'C:\Users\Tuhin\OneDrive\Desktop\DA_All\Python_dataquest\Part1 - Project 2\hacker_news.csv', encoding='utf-8') as file:
    read_file = reader(file)
    app_data  = list(read_file)
app_data[:2]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '08-04-2016 11:52']]

Print header

In [3]:
header = app_data[:1]
header

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]

In [4]:
hn = app_data[1:]

Display the first 5 rows of hn

In [5]:
hn[:6]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '08-04-2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12'],
 ['10482257',
  'Title II k

In [6]:
ask_posts,show_posts,other_posts= [],[],[]
for i in hn:
    title = i[1]
    check = title.lower()
    if check.startswith('ask hn'):
        ask_posts.append(i)
    elif check.startswith('show hn'):
        show_posts.append(i)
    else:
        other_posts.append(i)
print(len(ask_posts),len(show_posts),len(other_posts))

1744 1162 17194


First 2 rows of ask_post

In [7]:
ask_posts[:3]

[['12296411',
  'Ask HN: How to improve my personal website?',
  '',
  '2',
  '6',
  'ahmedbaracat',
  '8/16/2016 9:55'],
 ['10610020',
  'Ask HN: Am I the only one outraged by Twitter shutting down share counts?',
  '',
  '28',
  '29',
  'tkfx',
  '11/22/2015 13:43'],
 ['11610310',
  'Ask HN: Aby recent changes to CSS that broke mobile?',
  '',
  '1',
  '1',
  'polskibus',
  '05-02-2016 10:14']]

In [8]:
total_ask_comments = 0
total = 0
for i in ask_posts:
    comments = int(i[4])
    total_ask_comments += comments
    total+=1
avg_ask_comments = total_ask_comments/total 
avg_ask_comments

14.038417431192661

In [9]:
total_show_comments = 0
totals = 0
for i in show_posts:
    comments = int(i[4])
    total_show_comments += comments
    totals +=1
avg_show_comments = total_show_comments / totals
avg_show_comments

10.31669535283993

We can see on an average ask posts get more comments than show posts
The remaining analysis we'll continue in ask_posts

we'll determine if ask posts created at a certain time are more likely to attract comments.

In [10]:
import datetime as dt 
result_list  = []
for i in ask_posts:
    time = i[6]
    comments = int(i[4])
    result_list.append([time, comments])

In [11]:
result_list[:2]

[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29]]

Process of extracting hour

In [12]:
x= '8/16/2016 9:55'
print(dt.datetime.strptime(x,'%m/%d/%Y %H:%M').strftime('%H'))

09


In [13]:
counts_by_hour, comments_by_hour = {},{}
dateformat = '%m/%d/%Y %H:%M'
for i in result_list:
    date  =  i[0].replace('-','/')
    time = dt.datetime.strptime(date,dateformat).strftime('%H')
    comments = i[1]
    if time in counts_by_hour:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comments 
    else:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comments 
comments_by_hour

{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

Show the dict in key sorted manner

In [14]:
for i in sorted(comments_by_hour):
    print(i,'->', comments_by_hour[i])


00 -> 447
01 -> 683
02 -> 1381
03 -> 421
04 -> 337
05 -> 464
06 -> 397
07 -> 267
08 -> 492
09 -> 251
10 -> 793
11 -> 641
12 -> 687
13 -> 1253
14 -> 1416
15 -> 4477
16 -> 1814
17 -> 1146
18 -> 1439
19 -> 1188
20 -> 1722
21 -> 1745
22 -> 479
23 -> 543


In Value sorted

In [15]:
sorted_item = sorted(comments_by_hour.items(),key = lambda item:item[1], reverse=True)
for i in sorted_item:
    print(i[0],'->',i[1])

15 -> 4477
16 -> 1814
21 -> 1745
20 -> 1722
18 -> 1439
14 -> 1416
02 -> 1381
13 -> 1253
19 -> 1188
17 -> 1146
10 -> 793
12 -> 687
01 -> 683
11 -> 641
23 -> 543
08 -> 492
22 -> 479
05 -> 464
00 -> 447
03 -> 421
06 -> 397
04 -> 337
07 -> 267
09 -> 251


In [16]:
counts_by_hour

{'09': 45,
 '13': 85,
 '10': 59,
 '14': 107,
 '16': 108,
 '23': 68,
 '12': 73,
 '17': 100,
 '15': 116,
 '21': 109,
 '20': 80,
 '02': 58,
 '18': 109,
 '03': 54,
 '05': 46,
 '19': 110,
 '01': 60,
 '22': 71,
 '08': 48,
 '04': 47,
 '00': 55,
 '06': 44,
 '07': 34,
 '11': 58}

Calculate the average number of comments per post for posts created during each hour of the day

In [17]:
avg_hours = []
for i in comments_by_hour.keys():
    avg = comments_by_hour[i]/counts_by_hour[i]
    avg_hours.append([i,avg])
avg_hours = sorted(avg_hours)
avg_hours

[['00', 8.127272727272727],
 ['01', 11.383333333333333],
 ['02', 23.810344827586206],
 ['03', 7.796296296296297],
 ['04', 7.170212765957447],
 ['05', 10.08695652173913],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['08', 10.25],
 ['09', 5.5777777777777775],
 ['10', 13.440677966101696],
 ['11', 11.051724137931034],
 ['12', 9.41095890410959],
 ['13', 14.741176470588234],
 ['14', 13.233644859813085],
 ['15', 38.5948275862069],
 ['16', 16.796296296296298],
 ['17', 11.46],
 ['18', 13.20183486238532],
 ['19', 10.8],
 ['20', 21.525],
 ['21', 16.009174311926607],
 ['22', 6.746478873239437],
 ['23', 7.985294117647059]]

Sort by avg descending

In [18]:
sorted_avg_hours = sorted(avg_hours, key = lambda item: item[1], reverse = True )
sorted_avg_hours

[['15', 38.5948275862069],
 ['02', 23.810344827586206],
 ['20', 21.525],
 ['16', 16.796296296296298],
 ['21', 16.009174311926607],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['18', 13.20183486238532],
 ['17', 11.46],
 ['01', 11.383333333333333],
 ['11', 11.051724137931034],
 ['19', 10.8],
 ['08', 10.25],
 ['05', 10.08695652173913],
 ['12', 9.41095890410959],
 ['06', 9.022727272727273],
 ['00', 8.127272727272727],
 ['23', 7.985294117647059],
 ['07', 7.852941176470588],
 ['03', 7.796296296296297],
 ['04', 7.170212765957447],
 ['22', 6.746478873239437],
 ['09', 5.5777777777777775]]

In [19]:
# or ,

for i in sorted_avg_hours:
    print(i[1],'->', i[0])

38.5948275862069 -> 15
23.810344827586206 -> 02
21.525 -> 20
16.796296296296298 -> 16
16.009174311926607 -> 21
14.741176470588234 -> 13
13.440677966101696 -> 10
13.233644859813085 -> 14
13.20183486238532 -> 18
11.46 -> 17
11.383333333333333 -> 01
11.051724137931034 -> 11
10.8 -> 19
10.25 -> 08
10.08695652173913 -> 05
9.41095890410959 -> 12
9.022727272727273 -> 06
8.127272727272727 -> 00
7.985294117647059 -> 23
7.852941176470588 -> 07
7.796296296296297 -> 03
7.170212765957447 -> 04
6.746478873239437 -> 22
5.5777777777777775 -> 09


Top 5 Hours for Ask Posts Comments

In [28]:
for i in sorted_avg_hours[:5]:
    print(dt.datetime.strptime(i[0],'%H').strftime('%H:%M'),
          '->', "{:.2f}".format(i[1]),'average comments per post')

15:00 -> 38.59 average comments per post
02:00 -> 23.81 average comments per post
20:00 -> 21.52 average comments per post
16:00 -> 16.80 average comments per post
21:00 -> 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post